### ISSUE TITLE
websocket.read_message() never returns

### ISSUE THREAD

**Issue by dten** [snapview/tungstenite-rs#15]
hi. the slack crate uses this crate for websockets. Every now and then I get stuck in read_message and it never returns.
Either that or it returns a binary message (which we ignore) https://github.com/slack-rs/slack-rs/blob/master/src/lib.rs#L205

The server end of the system thinks there is no one connected though so it seems to me that we're in some zombie state. Is this expected to be possible? We used to used a different websocket-rs and had to set a manual timeout on the underlying socket. Is this someone you would expect we have to do? I was expecting ping/pong to take care of this but obviously it's not

Stacktrace below shows waiting in read

```
#0  0x00007fa984221102 in recv () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007fa9855ad42f in _$LT$std..net..tcp..TcpStream$u20$as$u20$std..io..Read$GT$::read::h08a6bc50c3293175 ()
    at /checkout/src/libstd/sys/unix/net.rs:161
#2  0x00007fa9854e3053 in std::panicking::try::do_call::h569c007b8386635e ()
#3  0x00007fa9855bd2cb in __rust_maybe_catch_panic () at /checkout/src/libpanic_unwind/lib.rs:98
#4  0x00007fa9854e3ad6 in openssl::ssl::bio::bread::h20a49d7450378b36 ()
#5  0x00007fa98491c139 in BIO_read () from /lib/x86_64-linux-gnu/libcrypto.so.1.0.0
#6  0x00007fa984c391e1 in ?? () from /lib/x86_64-linux-gnu/libssl.so.1.0.0
#7  0x00007fa984c3a41d in ?? () from /lib/x86_64-linux-gnu/libssl.so.1.0.0
#8  0x00007fa984c375e0 in ?? () from /lib/x86_64-linux-gnu/libssl.so.1.0.0
#9  0x00007fa985139667 in tungstenite::input_buffer::InputBuffer::read_from::h95abdbaba330b42c ()
#10 0x00007fa98515dc73 in slack::RtmClient::run::hcef610b3e818eb3e ()
#11 0x00007fa9852042d5 in sirbotalot::main::hfa1e56b4aa13694e ()
#12 0x00007fa9855b5fe6 in std::panicking::try::do_call::h689a21caeeef92aa () at /checkout/src/libstd/panicking.rs:454
#13 0x00007fa9855bd2cb in __rust_maybe_catch_panic () at /checkout/src/libpanic_unwind/lib.rs:98
#14 0x00007fa9855b6a97 in std::rt::lang_start::hf63d494cb7dd034c () at /checkout/src/libstd/panicking.rs:433
#15 0x00007fa983c5c7ed in __libc_start_main () from /lib/x86_64-linux-gnu/libc.so.6
#16 0x00007fa985134f1d in _start ()
```

**Comment by application-developer-DA** [comment]
> We used to used a different websocket-rs and had to set a manual timeout on the underlying socket. Is this someone you would expect we have to do?

Regardless of timeouts, `read_message()` would have returned if the connection had dropped, i.e. `read_message()` "unblocks" and returns the value if when we receive a websocket message **or** if some error occurs.

I would assume that it's blocked just because there are no messages coming from the server, but then I've noticed that you mentioned: _"The server end of the system thinks there is no one connected though"_. Hm... The thing is, if the connection is dropped by the server, `read_message()` would unblock and return an error.

Are you sure that `read_message()` is the function which blocks? The stack trace mentions `InputBuffer`, but it is also used when we're doing a WebSocket handshake. Could it be that you're blocked during the WebSocket handshake? I'm asking, because there was an issue which has been fixed in a recent version (https://github.com/snapview/tokio-tungstenite/issues/12). Could it be the case?

**Comment by agalakhov** [comment]
Hi,
tungstenite relies on the same socket timeout as well. If the OS does not report the connection to be dropped, which may be the case if the peer drops the connection without performing shutdown() first, tungstenite will wait until the OS reports an error. The default OS timeout may vary from 30 seconds to 30 minutes and more. Configure a reasonable socket timeout (should tungstenite do that for you?) and avoid closing without proper shutdown.

In general, treat read_message() just like TcpStream::read_exact(). They are very similar in their behavior.

**Comment by dten** [comment]
thanks for quick response. I've been using 0.2.4 since a week ago but I'm not 100% certain it's stuck not stuck at the handshake. i do log on_connect and on_open and there has not been either since the 12th so i'm fairly certain it's read_message.
i take it a client would never disconnect from a lack of a ping/pong ?

**Comment by agalakhov** [comment]
A client won't do anything from a lack of ping/pong, but the operating system may do.

**Comment by dten** [comment]
Well we're seeing a timeout as that seems to be an expected thing to do
